### ISSUE TITLE
CUDA stream synchronization issue in ModelRunnerCPP

### ISSUE THREAD

**Issue by xsxszab** [NVIDIA/TensorRT-LLM#6425]
### System Info

- This is a general, GPU-architecture-agnostic synchronization issue, confirmed to be reproducible on Ampere and Hopper architectures.

### Who can help?

This issue has been confirmed and addressed via discussions with @ruoqianguo. As per @ruoqianguo's recommendation, we would appreciate it if @dcampora and @syuoni could take a look and double check the issue.

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

The issue arises when using the `ModelRunnerCPP` to perform multimodal inference.

1. Load a LLM (e.g., LlamaForCausalLM) via `ModelRunnerCpp.from_dir`.
2. Call the model runner's `generate()` method, where the multimodal embeddings are passed to the engine via prompt tuning arguments.
3. If the input tensor and the model have different data type (e.g., Input type is fp16 while the models's type is bf16), the model will produce wrong logits (e.g., all elements are `nan`) at ~5% probability.

Example code:
```python
import tensorrt_llm
import torch
from tensorrt_llm.runtime import ModelRunnerCpp

runner_kwargs = dict(engine_dir=ENGINE_PATH, rank=tensorrt_llm.mpi_rank())
llm_engine = ModelRunnerCpp.from_dir(**runner_kwargs)

# Load example input embeddings
# All inputs have shape [batch_size, seq_len, hidden_size (same for all inputs)]
input_embeddings = []
for i in range(10):
      embedding = torch.load(f'{INPUT_EMBED_PREFIX}{i}.pt')
      input_embeddings.append(embedding)

def forward(input_tensor):
    batch_size, seq_len, hidden_size = input_tensor.shape
    # the LLM's vocab size is 250112, prompt tuning ids should be larger than or equal to this value
    batch_input_ids = torch.arange(250112, 250112 + seq_len)
    batch_input_ids = batch_input_ids.repeat(batch_size, 1)
    prompt_tasks = ",".join(np.arange(batch_size).astype(str))

    torch.cuda.current_stream().synchronize()
    res = llm_engine.generate(
        batch_input_ids=batch_input_ids,
        pad_id=1,
        end_id=2,
        max_new_tokens=1,
        normalize_log_probs=False,
        return_dict=True,
        prompt_tasks=prompt_tasks,
        prompt_table=input_tensor
    )
    logits = res['generation_logits'].flatten()
    probs = torch.nn.functional.softmax(logits, dim=-1)
    return probs

def test_func():
    # Each thread runs on a separate cuda stream
    with torch.cuda.stream(torch.cuda.Stream()):
          for input_tensor in input_embeddings:
              probs = forward(input_tensor)
              print(probs)


threads = []
start_time = time.time()
for i in range(NUM_THREADS):
    t = threading.Thread(target=test_func)
    threads.append(t)
    t.start()
for t in threads:
    t.join()
```

### Expected behavior

The model should produce correct output logits.

### actual behavior

The model produces wrong logits (e.g., all elements are `nan`) at ~5% probability when the `generate()` method is called concurrently by multiple threads. Theoretically, the bug exists even if there is only one thread accessing the LLM, but currently we only observe issues under concurrent access scenarios (>4 threads).

### additional notes

The root cause of this issue has been identified: the `generate()` method calls `_prepare_embedding_table` internally, which performs type conversion on the input prompt table (i.e., the input tensor in the example code). This is an asynchronous operation, however there's no stream sync after it to ensure the operation is completed. Therefore, the input data enqueued could be corrupted, leading to wrong output logits.

https://github.com/NVIDIA/TensorRT-LLM/blob/2d21bca25e60a6d9cfe31903c75cde29d6111145/tensorrt_llm/runtime/model_runner.py#L464-L474

This issue could be fixed by adding `torch.cuda.current_stream().synchronize()` after line 472 (confirmed via experiments on A30, A100 and H100).
Tracking PR: #6426

**Comment by karljang** [comment]
Thanks for your contribution. Closing since your PR was merged.
